### PR TITLE
Post-SMTB cleanup

### PR DESCRIPTION
**Pull request by ilsenatorov**
Most important changes:

* Added unit testing with pytest
* Added github action to run the tests
* Cleaned up redundant code
* Rearranged some files here and there
* Improved config/training logic for finetuning

### CODE REVIEW

File: smtb/tests/test_finetune.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        seed=42,
+    )
+
+
+def test_train(sample_config, mock_data_dir):
+    os.environ["WANDB_MODE"] = "dryrun"
+    data_dir = mock_data_dir
+    sample_config.dataset_path = data_dir
+
+    train(sample_config)
```
Comment (Old-Shatterhand) [inline smtb/tests/test_finetune.py:48]: Do we want to check something here or just check for failure?

**Review by Old-Shatterhand — CHANGES_REQUESTED** [review]
Applies to all files: Comment the code. It will be a project with >5 people working on and maintaining it, we need to set some documentation standards to not get completely lost before christmas.

File: smtb/tests/test_finetune.py (hunk 1 above)
Comment (ilsenatorov) [reply smtb/tests/test_finetune.py:48]: Just check that it runs for 2 epochs without crashing - there is probably a way to test this in a more sensible way though

File: .github/workflows/run_test.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        with:
+          path: ~/.cache/pip  # Cache pip packages
+          key: ${{ runner.os }}-pip-${{ matrix.python-version }}-${{ hashFiles('**/requirements.txt') }}
+          restore-keys: |
+            ${{ runner.os }}-pip-${{ matrix.python-version }}-
+
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
```
[the remark this answers is not in the document]
Comment (ilsenatorov) [reply .github/workflows/run_test.yaml:31]: AFAIK it's just common standard to do it this way, so it fails closer to the actual source. I don't feel super strong about this

File: smtb/model.py [hunk 3]
```diff
@@ -66,18 +29,45 @@ def test_step(self, batch: tuple[torch.Tensor, torch.Tensor]) -> torch.Tensor:
...
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return self.model(x).squeeze(1)
+
+    def shared_step(self, batch: tuple[torch.Tensor, torch.Tensor], name: str = "train") -> torch.Tensor:
+        x, y = batch
+        y_pred = self.forward(x).float()
+        y = y.float()
+        loss = F.mse_loss(y_pred, y)
+        self.log(f"{name}/loss", loss)
+        self.log(f"{name}/r2", M.functional.r2_score(y_pred, y))
```
[the remark this answers is not in the document]
Comment (ilsenatorov) [reply smtb/model.py:69]: Can be done, but I wouldn't focus on fixing that in this specific pr

File: .github/workflows/run_test.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Old-Shatterhand) [reply .github/workflows/run_test.yaml:31]: Can do. In the end, the requirements will always be there, and it doesn't matter. But if they weren't there, I suspect getting an error later at the pytest stage because some dependencies are not installed and not because the requirements file is non-existent.

**Comment by ilsenatorov** [comment]
> Applies to all files: Comment the code. It will be a project with >5 people working on and maintaining it, we need to set some documentation standards to not get completely lost before christmas.

I would suggest comments/docstring on code that is not self-explanatory. For example:

```python
def setup(self, stage: str | None = None):
        """Create train, val, test datasets."""
        self.train = DownstreamDataset(self.data_dir / "train", self.layer_num)
        self.valid = DownstreamDataset(self.data_dir / "valid", self.layer_num)
        self.test = DownstreamDataset(self.data_dir / "test", self.layer_num)

    def _get_dataloader(self, dataset: DownstreamDataset, shuffle: bool = False) -> torch.utils.data.DataLoader:
        """Create a DataLoader for a given dataset."""
        return DataLoader(
            dataset, batch_size=self.batch_size, num_workers=self.num_workers, shuffle=shuffle, collate_fn=collate_fn
        )

    def train_dataloader(self) -> DataLoader:
        return self._get_dataloader(self.train, shuffle=True)
```

The first 2 functions are not super obvious, so some comments are necessary.

The last one is pretty self-explanatory and requires no documentation (IMO).


But since we are in a phase of active development, I wouldn't enforce huge docstrings on every function. Type hints and good variable names should be enough for 95% of the code.

File: smtb/tests/test_data.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    train_loader = data_module.train_dataloader()
+    batch = next(iter(train_loader))
+    embeddings, labels = batch
+    assert embeddings.size(0) == 2
+    assert labels.size(0) == 2
+    assert embeddings.size(2) == 320
+    assert 50 <= embeddings.size(1) <= 60
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (ilsenatorov) [reply smtb/tests/test_data.py:40]: copilot

File: smtb/model.py
Comment (sdkaraban) [inline smtb/model.py:1]: are we going to add those classifications as models?

File: smtb/pooling.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        self.linear_key = nn.Linear(self.config.hidden_dim, 1)  # Project to a single dimension
+        self.linear_query = nn.Linear(self.config.hidden_dim, 1)
+        self.softmax = nn.Softmax(dim=-1)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        keys = self.linear_key(x)
+        queries = self.linear_query(x)
+        attention_scores = torch.matmul(queries, keys.transpose(-2, -1))
+        attention_weights = self.softmax(attention_scores)
+        pooled_output = torch.matmul(attention_weights, x).sum(dim=1)
```
Comment (sdkaraban) [inline smtb/pooling.py:38]: maybe add normalization here?

**Review by sdkaraban — APPROVED** [review]

File: .github/workflows/run_test.yaml [hunk 6]
```diff
@@ -0,0 +1,39 @@
+name: Pytest
+
+on: [push]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ["3.11"]
```
Comment (devium335) [inline .github/workflows/run_test.yaml:10]: Are we just testing for 3.11? Maybe add other versions like 3.10 or eeven 3.9.

File: requirements.txt [hunk 7]
```diff
@@ -1,11 +1,13 @@
...
 fair-esm
-jsonargparse
 wandb
 tokenizers
 transformers
 beartype
 datasets
 transformers[torch]
+pytest
+pytest-cov
```
Comment (devium335) [inline requirements.txt:13]: Should we start specifying versions to avoid breaking changes and use dependabot to update andreview version changes?

File: setup.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+setup(
+    name="smtb",
+    version="0.1.0",
+    author="Ilya Senatorov",
+    author_email="il.senatorov@protonmail.com",
+    url="https://github.com/kalininalab/SMTB2024",
+    packages=["smtb"],
+    install_requires=requirements,
+    python_requires=">=3.11",
```
Comment (devium335) [inline setup.py:15]: Again is 3.11 maybe too high, maybe 3.10?

File: requirements.txt (hunk 7 above)
Comment (ilsenatorov) [reply requirements.txt:13]: yeah, sure

File: setup.py (hunk 8 above)
Comment (ilsenatorov) [reply setup.py:15]: agreed

Comment (Old-Shatterhand) [reply setup.py:15]: I don't think, 3.11 is too high. I have another project with similar dependencies that is run on 3.11 and it works fine, so, I wouldn't expect any problems. But we can downgrade to 3.10 if you want.

File: .github/workflows/run_test.yaml (hunk 6 above)
Comment (Old-Shatterhand) [reply .github/workflows/run_test.yaml:10]: Related to [this comment](https://github.com/kalininalab/SMTB2024/pull/31#discussion_r1739983218). I'd test all Python versions that are actively developed, such as 3.12. But if we fix versions in the requirements, I don't see why we should test multiple versions. The installed packages will have the same versions across different tested Python versions.

File: setup.py (hunk 8 above)
Comment (danielkorkin) [reply setup.py]: Yes, maybe `>=3.10`

File: requirements.txt (hunk 7 above)
Comment (danielkorkin) [reply requirements.txt]: This should be easily accomplished by installing requirements and then running `pip freeze`
